### PR TITLE
removal of ContextAnnotationMapper and deprecation of ContextBinding

### LINKED ISSUE #146: Need support for @Context to inject SecurityContext
### Feature description

SecurityContext is an injectable interface that provides access to security related information.

In JAX-RS / jersey implementation, we use @Context to inject SecurityContext directly into the resource method as a parameter. And we refer to it later to fetch the user details from it.

    @POST
    public Response createDatabase(
            @NotNull DBRequest body,
            @Context SecurityContext securityContext)
    {

         :
         : <snip>
         :
         String user = securityContext.getUserPrincipal().getName();
         :
         :
    }

Though @Context is available in Micronaut but it seems to be different compared to @Context in JAX-RS.

Need support for @Context and SecurityContext in Micronaut with Micronaut Controller.

### PR DESCRIPTION
**Pull request by graemerocher**
Fixes #146

### CODE REVIEW

**Comment by sdelamo** [comment]
I will check this tomorrow.

**Review by sdelamo — CHANGES_REQUESTED** [review]
This is a breaking change

**Review by sdelamo — COMMENTED** [review]
I've extracted and merged the security PR out of the original one: 

- [Integrate JAX-RS Security Context with Micronaut Security](https://github.com/micronaut-projects/micronaut-jaxrs/pull/152)

Moreover, I have extracted this PR [Imporve ContextAnnotationBinder](https://github.com/micronaut-projects/micronaut-jaxrs/pull/153) but I think it lacks a test. 

With [Fix Incremental compilation](https://github.com/micronaut-projects/micronaut-jaxrs/pull/156) I am not able to build locally. 

The remainder of changes in this PR, e.g. `ContextAnnotationMapper` removal and the changes to the `ContextAnnotationBinder` class signature look like a breaking change.  Not sure if those classes are consider internal but they were not annotated with `@Internal`.

File: jaxrs-server-security/src/test/java/io/micronaut/jaxrs/security/bind/SimpleSecurityContextTest.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+        assertFalse(result.isEmpty());
+        assertEquals("fred", result.get("principal"));
+        assertEquals("false", result.get("secure"));
+        assertEquals("true", result.get("hasRole"));
+    }
+
+
+    @Override
+    public Map<String, String> getProperties() {
```
[the remark this answers is not in the document]
Comment (sdelamo) [reply jaxrs-server-security/src/test/java/io/micronaut/jaxrs/security/bind/SimpleSecurityContextTest.java]: I removed it at [Integrate JAX-RS Security Context with Micronaut Security](https://github.com/micronaut-projects/micronaut-jaxrs/pull/152). basic auth is enabled by default.

File: jaxrs-server/src/main/java/io/micronaut/jaxrs/runtime/ext/bind/ContextAnnotationBinder.java [hunk 2]
```diff
@@ -31,25 +36,37 @@
...
     private final BeanContext beanContext;
+    private final SimpleSecurityContextBinder securityBinder;
 
     /**
      * Default constructor.
      * @param beanContext The bean context
+     * @param simpleSecurityContextBinder The security context binder
      */
-    protected ContextAnnotationBinder(BeanContext beanContext) {
+    protected ContextAnnotationBinder(
```
[the remark this answers is not in the document]
Comment (sdelamo) [reply jaxrs-server/src/main/java/io/micronaut/jaxrs/runtime/ext/bind/ContextAnnotationBinder.java]: I created a new constructor, annotate it with `@Inject` and left the old one annotated with `@Deprecated` at [Integrate JAX-RS Security Context with Micronaut Security](https://github.com/micronaut-projects/micronaut-jaxrs/pull/152).

**Comment by graemerocher** [comment]
these are internal classes, since they are part of the compiler plugin not the public API

**Pull request merged**